### ISSUE TITLE
feat(cuda): cuDNN backward workspace allocation + convolution algorithm auto-tuning

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#220]
## Motivation

cuDNN convolution is the most compute-intensive operation in CNN training. Currently two performance issues leave significant speedup on the table:

1. **Backward passes run without workspace** — `CudaOps.java:745-747` hardcodes `wsSize = 0` and `workspace = NULL`, forcing cuDNN to fall back to the slowest possible algorithm variant. The forward pass already correctly queries and allocates workspace (line 673-674), so the backward pass is inconsistent.

2. **All passes hardcode `algo = 0` (IMPLICIT_GEMM)** — cuDNN offers significantly faster algorithms depending on kernel geometry (Winograd for 3×3, FFT-tiling for larger kernels, implicit precomp GEMM as a general fast default). PyTorch exposes this via `torch.backends.cudnn.benchmark = True`.

Supersedes #215 and #217, which addressed these problems independently.

---

### Forward pass — `CudaOps.java:631-692` ✅
```java
// Line 672-674: workspace correctly allocated
int algo = 0;                                            // hardcoded IMPLICIT_GEMM
long wsSize = cudnn.getConvolutionForwardWorkspaceSize(..., algo);
var workspace = pool.allocate(wsSize);                   // ✅ via GpuMemoryPool
```

### Backward data — `CudaOps.java:745-751` ❌
```java
// Line 746-747: workspace MISSING
long wsSize = 0;                                         // ❌ hardcoded zero
var workspace = MemorySegment.NULL;                      // ❌ no allocation
cudnn.convolutionBackwardData(..., wsSize, workspace);   // forced slowest path
```

### Backward filter — `CudaOps.java:753-756` ❌
```java
// Reuses the zero workspace from backward data              ❌ same problem
cudnn.convolutionBackwardFilter(..., wsSize, workspace);
```

### CuDnn.java bindings
| Function | Bound? | Public wrapper? | Used in CudaOps? |
|----------|--------|----------------|-------------------|
| `cudnnGetConvolutionForwardWorkspaceSize` | ✅ | ✅ `getConvolutionForwardWorkspaceSize()` | ✅ forward |
| `cudnnGetConvolutionForwardAlgorithm_v7` | ✅ (line 154) | ❌ missing | ❌ not called |
| `cudnnGetConvolutionBackwardDataWorkspaceSize` | ❌ missing | ❌ | ❌ |
| `cudnnGetConvolutionBackwardFilterWorkspaceSize` | ❌ missing | ❌ | ❌ |
| `cudnnFindConvolutionBackwardDataAlgorithm_v7` | ❌ missing | ❌ | ❌ |
| `cudnnFindConvolutionBackwardFilterAlgorithm_v7` | ❌ missing | ❌ | ❌ |

---

### Phase 1: Fix backward workspace allocation (from #215)

**Goal**: Backward passes allocate workspace the same way forward already does.

#### 1a. Add missing Panama bindings in `CuDnn.java`

Add two new method handles in the static initializer block (~line 160):
- `cudnnGetConvolutionBackwardDataWorkspaceSize` — signature: `(handle, filterDesc, dyDesc, convDesc, dxDesc, algo) → sizeInBytes`
- `cudnnGetConvolutionBackwardFilterWorkspaceSize` — signature: `(handle, xDesc, dyDesc, convDesc, dwDesc, algo) → sizeInBytes`

Add corresponding public wrapper methods following the pattern of `getConvolutionForwardWorkspaceSize()` (line 378-390).

#### 1b. Allocate workspace in backward passes in `CudaOps.java`

Replace lines 745-756:
```java
// Backward data: query + allocate workspace
int bwdDataAlgo = 0;
long bwdDataWsSize = cudnn.getConvolutionBackwardDataWorkspaceSize(..., bwdDataAlgo);
var bwdDataWs = pool.allocate(bwdDataWsSize);
try {
    cudnn.convolutionBackwardData(..., bwdDataAlgo, bwdDataWs, bwdDataWsSize, ...);
} finally {
    pool.release(bwdDataWs, bwdDataWsSize);
}

// Backward filter: query + allocate workspace (independently!)
int bwdFilterAlgo = 0;
long bwdFilterWsSize = cudnn.getConvolutionBackwardFilterWorkspaceSize(..., bwdFilterAlgo);
var bwdFilterWs = pool.allocate(bwdFilterWsSize);
try {
    cudnn.convolutionBackwardFilter(..., bwdFilterAlgo, bwdFilterWs, bwdFilterWsSize, ...);
} finally {
    pool.release(bwdFilterWs, bwdFilterWsSize);
}
```

**Important**: Each backward direction needs its own workspace query and allocation — data grad and filter grad have different workspace size requirements.

#### Expected impact: 10–30% training speedup

---

### Phase 2: Algorithm auto-tuning (from #217)

**Goal**: Benchmark all cuDNN algorithms per convolution configuration and cache the fastest one. Opt-in via `NetworkBuilder.cudnnBenchmark(true)`.

#### 2a. Add remaining Panama bindings in `CuDnn.java`

- `cudnnFindConvolutionForwardAlgorithm_v7` — already bound (line 154), needs public wrapper method
- `cudnnFindConvolutionBackwardDataAlgorithm_v7` — needs binding + wrapper
- `cudnnFindConvolutionBackwardFilterAlgorithm_v7` — needs binding + wrapper

These `_v7` variants return `cudnnConvolutionFwdAlgoPerf_t` structs with timing info. Define a corresponding Java record or struct layout for the result.

#### 2b. Algorithm cache in `CudaOps.java`

```java
// Cache key: everything that affects algorithm choice
record ConvConfig(int[] inputShape, int[] filterShape, int stride, int padding, int dilation) {}

// Cache value: best algorithm + its workspace requirement
record AlgoChoice(int algo, long workspaceBytes) {}

// Three caches (forward, backward-data, backward-filter)
private final Map<ConvConfig, AlgoChoice> fwdAlgoCache = new ConcurrentHashMap<>();
private final Map<ConvConfig, AlgoChoice> bwdDataAlgoCache = new ConcurrentHashMap<>();
private final Map<ConvConfig, AlgoChoice> bwdFilterAlgoCache = new ConcurrentHashMap<>();
```

On first encounter of a new `ConvConfig`:
1. Call `cudnnFind*Algorithm_v7()` to benchmark all candidates
2. Pick the fastest with `status == CUDNN_STATUS_SUCCESS`
3. Cache `(algo, workspaceBytes)` — valid for the lifetime of the `CudaOps` instance
4. Subsequent calls with the same config hit the cache (O(1) lookup)

#### 2c. Public API in `NetworkBuilder.java`

```java
// Opt-in: benchmark cuDNN algorithms on first convolution call
public NetworkBuilder cudnnBenchmark(boolean enable) { ... }
```

Pass this flag through to `CudaOps` (via the `ComputeOps` interface or constructor parameter). When disabled (default), use `IMPLICIT_GEMM` with workspace (Phase 1 behavior). When enabled, benchmark and cache.

#### 2d. cuDNN algorithm constants

Define algorithm constants in `CuDnn.java` for readability:
```java
// Forward algorithms
public static final int CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM         = 0;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM = 1;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_GEMM                  = 2;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_DIRECT                = 3;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_FFT                   = 4;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_FFT_TILING            = 5;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD              = 6;
public static final int CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED     = 7;
// + backward data and filter equivalents
```

#### Expected impact: 2–5x speedup on convolution-heavy models (ResNet, VGG)

---

## Files to modify

| File | Changes |
|------|---------|
| `lib-cuda/.../CuDnn.java` | Add backward workspace size bindings, find-algorithm bindings, algorithm constants, public wrapper methods |
| `lib-cuda/.../CudaOps.java` | Backward workspace allocation, algorithm caching, benchmark flag support |
| `lib/.../NetworkBuilder.java` | `cudnnBenchmark(boolean)` API |
| `lib/.../ComputeOps.java` | Possibly add `setBenchmark(boolean)` or pass config via constructor |

## Testing strategy

- **Unit**: Verify backward workspace sizes are > 0 for standard conv configs (3×3, 5×5, 1×1)
- **Correctness**: Gradient check — compare backward results with/without workspace (values must be identical within float tolerance)
- **Memory**: Assert workspace is returned to `GpuMemoryPool` after each backward call (no VRAM leak)
- **Auto-tuning**: Verify cache populates on first call and subsequent calls are cache hits
- **Integration**: Run TinyImageNet example, measure wall-clock speedup vs. current main

## Acceptance criteria

- [ ] `CuDnn.java`: `getConvolutionBackwardDataWorkspaceSize()` and `getConvolutionBackwardFilterWorkspaceSize()` bound and working
- [ ] `CudaOps.java`: backward data and filter passes allocate workspace via `GpuMemoryPool`
- [ ] Workspace released in `finally` block — no VRAM leak under any code path
- [ ] Backward pass produces numerically identical results (workspace only changes speed, not output)
- [ ] `CuDnn.java`: `findConvolution{Forward,BackwardData,BackwardFilter}Algorithm_v7()` bound
- [ ] Algorithm cache keyed on `(inputShape, filterShape, stride, padding, dilation)`
- [ ] Cache is `ConcurrentHashMap` — thread-safe for multi-stream future use
- [ ] `NetworkBuilder.cudnnBenchmark(true)` opt-in API added
- [ ] Default behavior (no benchmark): IMPLICIT_GEMM + workspace (Phase 1 only)
- [ ] All existing CUDA tests pass (`./gradlew :lib-cuda:test`)
- [ ] Measurable speedup on TinyImageNet CNN example

## References

- [cuDNN API — cudnnGetConvolutionBackwardDataWorkspaceSize](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-cnn-library.html)
- [cuDNN API — cudnnFindConvolutionForwardAlgorithm_v7](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-cnn-library.html)
- PyTorch equivalent: [`torch.backends.cudnn.benchmark`](https://pytorch.org/docs/stable/backends.html#torch.backends.cudnn.benchmark)
- cuDNN Developer Guide: [Convolution](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html)

**Comment by Samyssmile** [comment]
Fixed in ca10cbd. Backward workspace allocation + algorithm auto-tuning with cuDNN _v7 API benchmark cache.
